File: Checking.py
import numpy as np
from SortingAlgorythm.Python.HeapSort import HeapSort
from SortingAlgorythm.Python.MergeSort import MergeSort
from SortingAlgorythm.Python.QuickSort import QuickSort
import time
import os
import sys


def read_file(filepath):
    with open(filepath, 'r') as f:
        content = f.read().split()
        
    if not content:
        return 0, []
        
    n = int(content[0])
    data = [float(x) for x in content[1:n+1]]
    return n, data

def NumpySortTime(data):
    arr = np.array(data)
    
    start = time.time()
    sorted_arr = np.sort(arr)
    end = time.time()
    
    return (end - start) * 1000.0

def HeapSortTime(data):
    arr = list(data)

    start = time.time()
    HeapSort(arr)
    end = time.time()
    
    return (end - start) * 1000.0 

def MergeSortTime(data):
    arr = list(data)

    start = time.time()
    MergeSort(arr)
    end = time.time()
    
    return (end - start) * 1000.0

def QuickSortTime(data):
    arr = list(data)
    sys.setrecursionlimit(max(3000, len(arr) + 1000))

    start = time.time()
    QuickSort(arr)
    end = time.time()
    
    return (end - start) * 1000.0

def main():
    print(f"{'Test ID':<10} | {'Numpy':<10} | {'Heap':<10} | {'Merge':<10} | {'Quick':<10}")
    print("-" * 65)


    for test_id in range(1, 11):
        filename = f"Data/test{test_id}.inp"
        
        n, data = read_file(filename)

        t_numpy = NumpySortTime(data)
        t_heap = HeapSortTime(data)
        t_merge = MergeSortTime(data)
        t_quick = QuickSortTime(data)
        
        print(f"TEST {test_id:<5} | {t_numpy:>8.2f}ms | {t_heap:>8.2f}ms | {t_merge:>8.2f}ms | {t_quick:>8.2f}ms")

if __name__ == "__main__":
    main()

File: Checking.cpp
#include <bits/stdc++.h>
#include "SortingAlgorythm/C++/MergeSort.h"
#include "SortingAlgorythm/C++/QuickSort.h"
#include "SortingAlgorythm/C++/HeapSort.h"

using namespace std;

int n;
float a[1000005];
float tmp[1000005];

void READFILE(string FILEPATH){
    ifstream FILE(FILEPATH);
    FILE >> n;
    for (int i = 0; i < n; i ++ )
        FILE >> a[i];
    FILE.close();
}

float MergeSortTime() {
    for (int i = 0; i < n; i++) tmp[i] = a[i];
    clock_t start = clock();
    MergeSort(tmp, n);
    clock_t end = clock();
    return (float)(end - start) * 1000.0 / CLOCKS_PER_SEC;
}

float QuickSortTime() {
    for (int i = 0; i < n; i++) tmp[i] = a[i];
    clock_t start = clock();
    QuickSort(tmp, n);
    clock_t end = clock();
    return (float)(end - start) * 1000.0 / CLOCKS_PER_SEC;
}

float HeapSortTime() {
    for (int i = 0; i < n; i++) tmp[i] = a[i];
    clock_t start = clock();
    HeapSort(tmp, n);
    clock_t end = clock();
    return (float)(end - start) * 1000.0 / CLOCKS_PER_SEC;
}

float CPPSortTime() {
    for (int i = 0; i < n; i++) tmp[i] = a[i];
    clock_t start = clock();
    sort(tmp, tmp + n);
    clock_t end = clock();
    return (float)(end - start) * 1000.0 / CLOCKS_PER_SEC;  
}

int main(){
    
    for (int Test = 1; Test <= 10; Test ++ ){
        READFILE("Data/test" + to_string(Test) + ".inp");
        cout << "TEST " << Test;
        cout <<"|HeapSort: " <<HeapSortTime() <<"ms";
        cout <<"|MergeSort: " << MergeSortTime() <<"ms";
        cout <<"|QuickSort: " << QuickSortTime() <<"ms";
        cout <<"|C++: "<<CPPSortTime()<<"ms";
        cout <<endl; 
    }
}
